File: Sol18_4Sum/Solution18.py
from typing import List

class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        res = []
        nums.sort()

        for i in range(len(nums) - 3):
            if i > 0 and nums[i] == nums[i - 1]:
                continue

            for j in range(i + 1, len(nums) - 2):
                if j > i + 1 and nums[j] == nums[j - 1]:
                    continue

                left = j + 1
                right = len(nums) - 1

                while left < right:
                    total = nums[i] + nums[j] + nums[left] + nums[right]

                    if total < target:
                        left += 1
                    elif total > target:
                        right -= 1
                    else:
                        res.append([nums[i], nums[j], nums[left], nums[right]])
                        left += 1
                        right -= 1

                        while left < right and nums[left] == nums[left - 1]:
                            left += 1

                        while left < right and nums[right] == nums[right + 1]:
                            right -= 1

        return res


# Examples

solution = Solution()

# Example 1
nums1 = [1, 0, -1, 0, -2, 2]
target1 = 0

print("Example 1")
print("Input:", nums1)
print("Target:", target1)
print("Output:", solution.fourSum(nums1, target1))

print()

# Example 2
nums2 = [2, 2, 2, 2, 2]
target2 = 8

print("Example 2")
print("Input:", nums2)
print("Target:", target2)
print("Output:", solution.fourSum(nums2, target2))

File: Sol18_4Sum/Solution18.java
import java.util.*;

class Solution {
    public List<List<Integer>> fourSum(int[] nums, int target) {
        List<List<Integer>> list = new ArrayList<>();

        Arrays.sort(nums);

        for (int i = 0; i < nums.length - 3; i++) {

            if (i > 0 && nums[i] == nums[i - 1]) {
                continue;
            }

            for (int j = i + 1; j < nums.length - 2; j++) {

                if (j > i + 1 && nums[j] == nums[j - 1]) {
                    continue;
                }

                int left = j + 1;
                int right = nums.length - 1;

                while (left < right) {

                    long sum = (long) nums[i]
                             + nums[j]
                             + nums[left]
                             + nums[right];

                    if (sum == target) {

                        list.add(Arrays.asList(
                            nums[i],
                            nums[j],
                            nums[left],
                            nums[right]
                        ));

                        left++;
                        right--;

                        while (left < right &&
                               nums[left] == nums[left - 1]) {
                            left++;
                        }

                        while (left < right &&
                               nums[right] == nums[right + 1]) {
                            right--;
                        }

                    } else if (sum < target) {
                        left++;
                    } else {
                        right--;
                    }
                }
            }
        }

        return list;
    }

    public static void main(String[] args) {

        Solution solution = new Solution();

        // Example 1
        int[] nums1 = {1, 0, -1, 0, -2, 2};
        int target1 = 0;

        System.out.println("Example 1");
        System.out.println("Input: nums = " + Arrays.toString(nums1));
        System.out.println("Target = " + target1);
        System.out.println("Output: " + solution.fourSum(nums1, target1));

        System.out.println();

        // Example 2
        int[] nums2 = {2, 2, 2, 2, 2};
        int target2 = 8;

        System.out.println("Example 2");
        System.out.println("Input: nums = " + Arrays.toString(nums2));
        System.out.println("Target = " + target2);
        System.out.println("Output: " + solution.fourSum(nums2, target2));
    }
}